app/services/user_timezone.rb

```diff
@@ -0,0 +1,43 @@
+class UserTimezone
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def wrong_timezone?
+    return false if user.timezone.blank? || user.detected_timezone.blank?
+
+    user.timezone != user.detected_timezone && user.ignored_timezone != user.detected_timezone
+  end
+
+  def check_user!(ip_address)
+    return if ip_address.blank? || local_ip_address?(ip_address)
+
+    timezone = detected_timezone(ip_address)
+    return if timezone == user.detected_timezone
+
+    user.update! detected_timezone: timezone
+  end
+
+  def update!
+    user.update! timezone: user.detected_timezone, ignored_timezone: nil
+  end
+
+  def ignore!
+    user.update! ignored_timezone: user.detected_timezone
+  end
+
+  private
+
+    def local_ip_address?(ip_address)
+      %w[127.0.0.1 ::1].include?(ip_address)
+    end
+
+    def detected_timezone(ip_address)
+      Rails.cache.fetch("ip_address_time_zone:#{ip_address}", expires_in: 1.week) do
```

Comment:
the 1 week expiry seems fine here since I doubt an IP address's location will even change at all